teos/src/tx_index.rs

```diff
@@ -0,0 +1,425 @@
+    tip: u32,
+    /// Maximum size of the index.
+    size: usize,
+}
+
+impl<K: std::fmt::Debug, V: std::fmt::Debug> fmt::Display for TxIndex<K, V> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "index: {:?}\n\nblocks: {:?}\n\ntx_in_block: {:?}\n\nsize: {}",
+            self.index, self.blocks, self.tx_in_block, self.size
+        )
+    }
+}
+
+pub trait Indexable<K, V>
+where
+    K: Key + std::cmp::Eq + Copy,
+    V: Value + Clone,
+    Self: Sized,
+{
+    /// Creates a new [Indexable].
+    fn new(last_n_blocks: &[ValidatedBlock], height: u32) -> Self;
+    /// Returns a reference to index.
+    fn index(&self) -> &HashMap<K, V>;
+    /// Returns a muable reference to index.
+    fn index_mut(&mut self) -> &mut HashMap<K, V>;
+    /// Returns a reference to blocks.
+    fn blocks(&self) -> &VecDeque<BlockHash>;
+    /// Returns a mutable reference to blocks.
+    fn blocks_mut(&mut self) -> &mut VecDeque<BlockHash>;
+    /// Returns a reference to tx_in_block.
+    fn tx_in_block(&self) -> &HashMap<BlockHash, Vec<K>>;
+    /// Returns a mutable reference to tx_in_block.
+    fn tx_in_block_mut(&mut self) -> &mut HashMap<BlockHash, Vec<K>>;
+    /// Returns the height of the last block covered by the cache.
+    fn tip(&self) -> u32;
+    /// Updates the height of the last block coveted by the cache.
+    fn update_tip(&mut self);
+    /// Returns the size of the cache.
+    fn size(&self) -> usize;
+
+    /// Gets an item from the index if present. [None] otherwise.
+    fn get<'a>(&'a self, k: &'a K) -> Option<&V> {
+        self.index().get(k)
+    }
+
+    /// Checks whether the index contains a certain key.
+    fn contains_key(&self, k: &K) -> bool {
+        self.index().contains_key(k)
+    }
+
+    /// Checks if the index if full.
+    fn is_full(&self) -> bool {
+        self.blocks().len() > self.size()
+    }
+
+    /// Get's the height of a given block based on its position in the block queue.
+    fn get_height(&self, block_hash: &BlockHash) -> Option<usize> {
+        let pos = self.blocks().iter().position(|x| x == block_hash)?;
+        Some(self.tip() as usize + pos + 1 - self.blocks().len())
+    }
+
+    /// Initializes the index with data from the last n known blocks.
+    ///
+    /// This is only meant to be used when creating the [TxIndex].
+    fn _init(&mut self, last_n_blocks: &[ValidatedBlock]) {
+        for block in last_n_blocks.into_iter().rev() {
+            if let Some(prev_block_hash) = self.blocks().back() {
+                if block.header.prev_blockhash != *prev_block_hash {
+                    panic!("last_n_blocks contains unchained blocks");
+                }
+            };
+
+            let map = block
+                .txdata
+                .iter()
+                .map(|tx| {
+                    (
+                        K::from_txid(tx.txid()),
+                        V::from_tx_and_header(tx, block.header),
```

Comment:
Yeah, I'm not a big fan either to be honest here. This is so that I could make the constructor part of the blanket implementation so I didn't need to implement it for both `Locator:Transaction` pair (`Watcher`) and the `Txid::BlockHash` map (`Responder`).

This has been updated in the last rebase though so it looks a bit more idiomatic, but I'm still not a big fan of it.